### PR TITLE
[Security] Allow switching to another user when already switched

### LINKED ISSUE #34690: [Security] Allow to switch user when already switching
**Description**  

The current way this feature works, an authenticated user already impersonating another needs to exit impersonation before switching to another user. Otherwise, he'll hit:

https://github.com/symfony/symfony/blob/2b97a0f5f1ffd2938f2a1a68f333ba9292a270d0/src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php#L134

First thing: I don't think a `LogicException` is the most suitable here, as it'll generate a 500.
Secondly, is there reasons to forbid this by default?

Then, what about allowing to configure this on the switch user listener? Different strategies might exist:
1. reject, with an appropriate 4XX exception
1. generate a response suggesting to exit & switch to the new user, out-of-the-box.
1. just switch, seamlessly.

Throwing a dedicated exception might allow to perform the second option in userland, but would miss an easy way to switch to the new user once exit.

**Example**  

- Could be a `switch_user` listener option:

```yml
security:
    firewalls:
        # [...]
        main:
            # [...]
            switch_user:
                already_switching_strategy: ~ # one of reject, confirm or switch; reject as default.
```

- a query parameter to force switching
- or the custom exception coupled with error handling in userland. Would miss a native way to exit+switch to the new user though.

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #34690
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml [hunk 1]
```diff
@@ -202,6 +202,7 @@
             <argument>ROLE_ALLOWED_TO_SWITCH</argument>
             <argument type="service" id="event_dispatcher" on-invalid="null"/>
             <argument>false</argument> <!-- Stateless -->
+            <argument>false</argument> <!-- Consecutive Switching -->
```
Comment (noniagriconomie) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml]:
```suggestion
            <argument>false</argument> <!-- Seamlessl Switching -->
```

as you seem to use this word on the changelog and comment :)

**Review by noniagriconomie — APPROVED** [review]
thx

File: src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php [hunk 2]
```diff
@@ -142,9 +143,15 @@ private function attemptSwitchUser(Request $request, string $username): ?TokenIn
         if (null !== $originalToken) {
             if ($token->getUsername() === $username) {
                 return $token;
+            } elseif (!$this->allowAlreadySwitched) {
+                $e = new AlreadySwitchedException(sprintf('You are already switched to "%s" user.', $token->getUsername()));
```
Comment (ogizanagi) [inline src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php]: What about adding this exception in 3.4 as a bug fix?

File: src/Symfony/Component/Security/Core/Exception/AlreadySwitchedException.php [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+namespace Symfony\Component\Security\Core\Exception;
+
+/**
+ * Thrown when trying to switch to another user while being already switched.
+ *
+ * @author Robin Chalas <robin.chalas@gmail.com>
+ */
+final class AlreadySwitchedException extends AuthenticationException
```
Comment (ogizanagi) [inline src/Symfony/Component/Security/Core/Exception/AlreadySwitchedException.php]: Perhaps we could extend `CustomUserMessageAuthenticationException` to provide a sensible & safe message to show to the user?

File: src/Symfony/Bundle/SecurityBundle/CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 -----
 
  * Added security configuration for priority-based access decision strategy
+ * Added `switch_user.allow_already_switched` option to allow switching seamlessly when already switched (default  `false`)
```
Comment (ogizanagi) [inline src/Symfony/Bundle/SecurityBundle/CHANGELOG.md]:
```suggestion
 * Added `switch_user.allow_already_switched` option to allow switching seamlessly when already switched (default `false`)
```

**Review by wouterj — APPROVED** [review]
Great, thanks!

File: src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php (hunk 2 above)
Comment (wouterj) [reply src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php]: I think this should be considered a feature and not a bug fix. (in the issue, we even argued whether or not this is a BC break)

**Comment by nicolas-grekas** [comment]
What would be the drawback of always allowing the switch? Adding a new option should be thought twice.

File: src/Symfony/Component/Security/Core/Exception/AlreadySwitchedException.php (hunk 3 above)
Comment (wouterj) [reply src/Symfony/Component/Security/Core/Exception/AlreadySwitchedException.php]: -1, we do not do that for any other core exception. If a user wants a more detailed message, they can do so by creating a new exception in an exception listener. (which is also why this specific exception is usefull: you can catch only this one)

**Comment by ogizanagi** [comment]
We've discussed a bit about this with @chalasr. It's not easy to provide a satisfying default out-of-the-box 
 to show an explanation about the error to the user anyway. I'd be in favor of always allowing the switch, and this could even be considered a bug fix to me.

File: src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php (hunk 2 above)
Comment (ogizanagi) [reply src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php]: I still don't see any BC about changing the exception here.

**Review by nicolas-grekas — APPROVED** [review]
Would work as a bug fix on 4.4 to me.

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/SwitchUserTest.php [hunk 5]
```diff
@@ -29,15 +29,15 @@ public function testSwitchUser($originalUser, $targetUser, $expectedUser, $expec
...
     {
         $client = $this->createAuthenticatedClient('user_can_switch');
 
         $client->request('GET', '/profile?_switch_user=user_cannot_switch_1');
         $client->request('GET', '/profile?_switch_user=user_cannot_switch_2');
 
-        $this->assertEquals(500, $client->getResponse()->getStatusCode());
-        $this->assertEquals('user_cannot_switch_1', $client->getProfile()->getCollector('security')->getUser());
+        $this->assertEquals(200, $client->getResponse()->getStatusCode());
+        $this->assertEquals('user_cannot_switch_2', $client->getProfile()->getCollector('security')->getUser());
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/SwitchUserTest.php]: this means the lowest version of security-http should be updated to make test green, isn't it?

Comment (chalasr) [reply src/Symfony/Bundle/SecurityBundle/Tests/Functional/SwitchUserTest.php]: Seems worse than that actually, SecurityBundle already requires security-http:^5.1. 
IIUC this test would need to be relaxed on a lower branch for this to merged on master.
Merging it as a bugfix should fix the problem.

**Comment by chalasr** [comment]
@nicolas-grekas Did you mean 3.4?

**Comment by nicolas-grekas** [comment]
3.4 then yes :)

**Comment by chalasr** [comment]
Rebased. Failing build expected with deps=high

**Review by ogizanagi — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @chalasr.

**Comment by noniagriconomie** [comment]
Thanks for merging it into 3.4

**Pull request merged**